### PR TITLE
python3Packages.watchdog: fix darwin-x86_64 build

### LINKED ISSUE #113777: pythonPackages.watchdog 2.0.0 is broken on darwin
**Describe the bug**

963299d8fc7afb7f0401b9e547b21236deab8bbc upgraded `pythonPackages.watchdog` to 2.0.0, but this version is broken on darwin:

```
...
creating build/temp.macosx-10.6-x86_64-3.8
creating build/temp.macosx-10.6-x86_64-3.8/src
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -DWATCHDOG_VERSION_STRING="2.0.0" -DWATCHDOG_VERSION_MAJOR=2 -DWATCHDOG_VERSION_MINOR=0 -DWATCHDOG_VERSION_BUILD=0 -I/nix/store/5vn5f6slfg1nw8q0bhapd3vgk9j7559i-python3-3.8.7/include/python3.8 -c src/watchdog_fsevents.c -o build/temp.macosx-10.6-x86_64-3.8/src/watchdog_fsevents.o -std=c99 -pedantic -Wall -Wextra -fPIC -Wno-nullability-completeness -Wno-nullability-extension -Wno-newline-eof -Wno-error=unused-command-line-argument
clang-7: warning: argument unused during compilation: '-fno-strict-overflow' [-Wunused-command-line-argument]
src/watchdog_fsevents.c:191:25: error: use of undeclared identifier 'kFSEventStreamEventFlagItemCloned'
FLAG_PROPERTY(IsCloned, kFSEventStreamEventFlagItemCloned)
                        ^
src/watchdog_fsevents.c:404:56: error: use of undeclared identifier 'kFSEventStreamEventExtendedDataPathKey'
        cf_path = CFDictionaryGetValue(path_info_dict, kFSEventStreamEventExtendedDataPathKey);
                                                       ^
src/watchdog_fsevents.c:405:57: error: use of undeclared identifier 'kFSEventStreamEventExtendedFileIDKey'
        cf_inode = CFDictionaryGetValue(path_info_dict, kFSEventStreamEventExtendedFileIDKey);
                                                        ^
src/watchdog_fsevents.c:585:40: error: use of undeclared identifier 'kFSEventStreamCreateFlagUseExtendedData'
                                     | kFSEventStreamCreateFlagUseExtendedData
                                       ^
4 errors generated.
error: command 'clang' failed with exit status 1
```

It appears this is because `nixpkgs`' macOS SDK is 10.12, but the symbols that cause problems in 2.0.0 are only available in 10.13+:

https://developer.apple.com/documentation/coreservices/kfseventstreameventextendeddatapathkey

### PR DESCRIPTION
**Pull request by cpcloud**
###### Description of changes

This change allows the `watchdog` package to build on x86_64-darwin.

It turns out that the use of the fsevents API isn't a hard requirement and
watchdog will fallback to a kqueue-based implementation of the equivalent
functionality on platforms that don't support fsevents or always, with the
included patch.

I don't have access to a macOS machine to test the functionality, so any help
there would be much appreciated.

Fixes: #113777
Closes: #156597

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by cpcloud** [comment]
@ofborg build python3Packages.watchdog

**Comment by cpcloud** [comment]
@ofborg build python3Packages.watchdog

**Comment by cpcloud** [comment]
@ofborg build python3Packages.watchdog

**Review by kubukoz — APPROVED** [review]

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 171388` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

**Comment by Mindavi** [comment]
ZHF: https://github.com/NixOS/nixpkgs/issues/172160.

Also, as Jon said, please do consider making this an upstream PR and just fetchpatch'ing it. Then hopefully there's no need to update this patch every time the package is updated. Upstream looks quite active so I bet they're willing to take a patch and maybe test it on a Darwin machine.

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 171388` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

**Comment by risicle** [comment]
On macos 10.15:

```
_______________________ test_separate_consecutive_moves ________________________

    @pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
    def test_separate_consecutive_moves():
        mkdir(p('dir1'))
        mkfile(p('dir1', 'a'))
        mkfile(p('b'))
        start_watching(p('dir1'))
        mv(p('dir1', 'a'), p('c'))
        mv(p('b'), p('dir1', 'd'))
    
        dir_modif = DirModifiedEvent(p('dir1'))
        a_deleted = FileDeletedEvent(p('dir1', 'a'))
        d_created = FileCreatedEvent(p('dir1', 'd'))
    
        expected_events = [a_deleted, dir_modif, d_created, dir_modif]
    
        if platform.is_windows():
            expected_events = [a_deleted, d_created]
    
        if platform.is_bsd() or platform.is_darwin():
            # Due to the way kqueue works, we can't really order
            # 'Created' and 'Deleted' events in time, so creation queues first
            expected_events = [d_created, a_deleted, dir_modif, dir_modif]
    
        for expected_event in expected_events:
>           expect_event(expected_event)

a_deleted  = <FileDeletedEvent: event_type=deleted, src_path='/private/tmp/nix-build-python3.9-watchdog-2.1.7.drv-0/tmpkj5_dju1/dir1/a', is_directory=False>
d_created  = <FileCreatedEvent: event_type=created, src_path='/private/tmp/nix-build-python3.9-watchdog-2.1.7.drv-0/tmpkj5_dju1/dir1/d', is_directory=False>
dir_modif  = <DirModifiedEvent: event_type=modified, src_path='/private/tmp/nix-build-python3.9-watchdog-2.1.7.drv-0/tmpkj5_dju1/dir1', is_directory=True>
expected_event = <FileCreatedEvent: event_type=created, src_path='/private/tmp/nix-build-python3.9-watchdog-2.1.7.drv-0/tmpkj5_dju1/dir1/d', is_directory=False>
expected_events = [<FileCreatedEvent: event_type=created, src_path='/private/tmp/nix-build-python3.9-watchdog-2.1.7.drv-0/tmpkj5_dju1/di...t_type=modified, src_path='/private/tmp/nix-build-python3.9-watchdog-2.1.7.drv-0/tmpkj5_dju1/dir1', is_directory=True>]

tests/test_emitter.py:343: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

expected_event = <FileCreatedEvent: event_type=created, src_path='/private/tmp/nix-build-python3.9-watchdog-2.1.7.drv-0/tmpkj5_dju1/dir1/d', is_directory=False>
timeout = 2

    def expect_event(expected_event, timeout=2):
        """ Utility function to wait up to `timeout` seconds for an `event_type` for `path` to show up in the queue.
    
        Provides some robustness for the otherwise flaky nature of asynchronous notifications.
        """
        try:
            event = event_queue.get(timeout=timeout)[0]
>           assert event == expected_event
E           AssertionError: assert <FileDeletedEvent: event_type=deleted, src_path='/private/tmp/nix-build-python3.9-watchdog-2.1.7.drv-0/tmpkj5_dju1/dir1/a', is_directory=False> == <FileCreatedEvent: event_type=created, src_path='/private/tmp/nix-build-python3.9-watchdog-2.1.7.drv-0/tmpkj5_dju1/dir1/d', is_directory=False>

event      = <FileDeletedEvent: event_type=deleted, src_path='/private/tmp/nix-build-python3.9-watchdog-2.1.7.drv-0/tmpkj5_dju1/dir1/a', is_directory=False>
expected_event = <FileCreatedEvent: event_type=created, src_path='/private/tmp/nix-build-python3.9-watchdog-2.1.7.drv-0/tmpkj5_dju1/dir1/d', is_directory=False>
timeout    = 2

tests/test_emitter.py:98: AssertionError
```

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 171388` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages marked as broken and skipped:</summary>
  <ul>
    <li>hydrus</li>
    <li>maestral</li>
    <li>python310Packages.Nikola</li>
    <li>python310Packages.maestral</li>
    <li>python39Packages.Nikola</li>
    <li>python39Packages.maestral</li>
  </ul>
</details>
<details>
  <summary>38 packages failed to build:</summary>
  <ul>
    <li>ablog</li>
    <li>aws-sam-cli</li>
    <li>chia</li>
    <li>coconut (python39Packages.coconut)</li>
    <li>hovercraft</li>
    <li>markdown-pp</li>
    <li>mkdocs (python39Packages.mkdocs)</li>
    <li>netbox</li>
    <li>octoprint</li>
    <li>paperless-ngx</li>
    <li>pympress</li>
    <li>python310Packages.bpython</li>
    <li>python310Packages.chalice</li>
    <li>python310Packages.coconut</li>
    <li>python310Packages.easywatch</li>
    <li>python310Packages.lektor</li>
    <li>python310Packages.mkdocs</li>
    <li>python310Packages.mkdocs-material</li>
    <li>python310Packages.mkdocs-minify</li>
    <li>python310Packages.mkdocs-redirects</li>
    <li>python310Packages.ndjson</li>
    <li>python310Packages.pytest-watch</li>
    <li>python310Packages.staticjinja</li>
    <li>python310Packages.watchdog</li>
    <li>python39Packages.bpython</li>
    <li>python39Packages.chalice</li>
    <li>python39Packages.easywatch</li>
    <li>python39Packages.lektor</li>
    <li>python39Packages.mkdocs-material</li>
    <li>python39Packages.mkdocs-minify</li>
    <li>python39Packages.mkdocs-redirects</li>
    <li>python39Packages.ndjson</li>
    <li>python39Packages.pytest-watch</li>
    <li>staticjinja (python39Packages.staticjinja)</li>
    <li>python39Packages.watchdog</li>
    <li>skjold</li>
    <li>streamlit</li>
    <li>topydo</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
> @ofborg build python3Packages.watchdog

Please edit your comments in the future and append a space to trigger ofborg again.

File: pkgs/development/python-modules/watchdog/default.nix [hunk 1]
```diff
@@ -20,6 +20,10 @@ buildPythonPackage rec {
     sha256 = "sha256-P9R4FTU76cRO68lMwo/iaysMW9iJ2vxKWny9+SQUNIA=";
   };
 
+  patches = lib.optionals (stdenv.isDarwin && !stdenv.isAarch64) [
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/watchdog/default.nix:23]: Can we make this unconditional?

**Comment by risicle** [comment]
Do we want to get this in for 22.05? It's possible my result is spurious and I don't think this _breaks_ anything AFAIK.

**Comment by cpcloud** [comment]
I won't have time to work on this for a week or two, and it sounds like there's some additional legwork to do, which in principle I agree should be done.

@jonringer I believe this patch is related to nix. Nix doesn't support the late(r|st) macos sdk which this package requires to compile on macos machines with the later sdk. The package works fine on older sdks without the patch.

**Comment by jmgilman** [comment]
May be unrelated, but I was unable to [get the tests to pass](https://github.com/NixOS/nixpkgs/issues/175032) on this package with darwin-aarch64 unless I explicitly disabled the fsevents test (among other things).

**Pull request merged**